core/src/main/scala/org/apache/spark/scheduler/TaskAssigner.scala

```diff
@@ -0,0 +1,226 @@
+    assigner
+  }
+}
+
+/**
+ * Assign the task to workers with available cores in roundrobin manner.
+ */
+class RoundRobinAssigner extends TaskAssigner {
+  private var currentOfferIndex = 0
+
+  override def construct(workOffer: Seq[WorkerOffer]): Unit = {
+    offer = Random.shuffle(workOffer.map(o => new OfferState(o)))
+  }
+
+  override def init(): Unit = {
+    currentOfferIndex = 0
+  }
+
+  override def hasNext: Boolean = currentOfferIndex < offer.size
+
+  override def next(): OfferState = {
+    offer(currentOfferIndex)
+  }
+
+  override def offerAccepted(assigned: Boolean): Unit = {
+    currentOfferIndex += 1
+  }
+}
+
+/**
+ * Assign the task to workers with the most available cores. It other words, BalancedAssigner tries
+ * to distribute the task across workers in a balanced way. Potentially, it may alleviate the
+ * workers' memory pressure as less tasks running on the same workers, which also indicates that
+ * the task itself can make use of more computation resources, e.g., hyper-thread, across clusters.
+ */
+class BalancedAssigner extends TaskAssigner {
+  implicit val ord: Ordering[OfferState] = new Ordering[OfferState] {
+    def compare(x: OfferState, y: OfferState): Int = {
+      return Ordering[Int].compare(x.coresAvailable, y.coresAvailable)
+    }
+  }
+  private val maxHeap: PriorityQueue[OfferState] = new PriorityQueue[OfferState]()
+  private var currentOffer: OfferState = _
+
+  override def construct(workOffer: Seq[WorkerOffer]): Unit = {
+    offer = Random.shuffle(workOffer.map(o => new OfferState(o)))
+  }
+
+  override def init(): Unit = {
+    maxHeap.clear()
+    offer.filter(_.coresAvailable >= cpuPerTask).foreach(maxHeap.enqueue(_))
+  }
+
+  override def hasNext: Boolean = maxHeap.nonEmpty
+
+  override def next(): OfferState = {
+    currentOffer = maxHeap.dequeue()
+    currentOffer
+  }
+
+  override def offerAccepted(assigned: Boolean): Unit = {
+    if (currentOffer.coresAvailable >= cpuPerTask && assigned) {
+      maxHeap.enqueue(currentOffer)
+    }
+  }
+}
+
+/**
+ * Assign the task to workers with the least available cores. In other words, PackedAssigner tries
+ * to schedule tasks to fewer workers.  As a result, there will be idle workers without any tasks
+ * assigned if more than required workers are reserved. If the dynamic allocator is enabled,
+ * these idle workers will be released by driver. The released resources can then be allocated to
+ * other jobs by underling resource manager. This assigner can potentially reduce the resource
+ * reservation for jobs, which over allocate resources than they need.
+ */
+class PackedAssigner extends TaskAssigner {
+  private var sorted: Seq[OfferState] = _
+  private var currentOfferIndex = 0
+  private var currentOffer: OfferState = _
+
+  override def init(): Unit = {
+    currentOfferIndex = 0
+    sorted = offer.filter(_.coresAvailable >= cpuPerTask).sortBy(_.coresAvailable)
```

Comment:
Sounds like the shuffling reason of `BalancedAssigner` can be applied here too? If no shuffling, this ensures that initial sorted offers will be randomized when cores are the same, right?